[EPM] Make API consistent for package installation and removal

skh opened the pull request:
## Summary

Implements https://github.com/elastic/kibana/issues/47724 by not allowing the optional `asset` path parameter on the routes for installation and removal of packages.

This PR changed directions while I was working on it: installation/removal of ingest pipelines will now happen during datasource handling (see https://github.com/elastic/kibana/issues/49032), so I removed the installation part during package installation. I kept the removal part but I'm not sure we'll need it at all.

So what's left over in this PR is mainly removing functionality and a few type changes. @jfsiii what do you think, are these changes worth doing now?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/5765/)
* Commit: fc737b1b5484ea0139ebc5699b0d4389f301e31d

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/6721/)
* Commit: 3842acb54ceebcae20fb2fdf70c33d911d52d441

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/6768/)
* Commit: 5a218a5677e4faaeed296cfa39661e38886742ff

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/6787/)
* Commit: 36699b5d489f0fbe9eaa8aa5b6759e51a8083529

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/7855/)
* Commit: 2df30c5972a3b6e529386737447cb65255db2eda

jfsiii commented:
@skh Yes, I think these are worth doing .LGTM overall. Please convert it to a standard PR and we can discuss.

CI is failing because `timelion-sheet` was removed as an `AssetType` in two of three referenced places. If the removal was intentional, can you add a note about why? It was included because it's listed as a `type` in the [Saved Objects API docs](https://www.elastic.co/guide/en/kibana/master/saved-objects-api-create.html#saved-objects-api-create-path-params)

jfsiii commented on x-pack/legacy/plugins/integrations_manager/server/packages/install.ts:
```diff
@@ -8,35 +8,40 @@ import { SavedObject, SavedObjectsClientContract } from 'src/core/server/';
+
+  // this is a for loop because using forEach with an async callback wouldn't wait for the
+  // awaits to resolve, and we'd return an empty result
+  for (let i = 0; i < assetTypes.length; i++) {
```
Can also write this as

```suggestion
  for (const assetType of assetTypes) {
```

jfsiii commented on x-pack/legacy/plugins/integrations_manager/server/packages/install.ts:
```diff
@@ -8,35 +8,40 @@ import { SavedObject, SavedObjectsClientContract } from 'src/core/server/';
+  let result = {};
+
+  // this is a for loop because using forEach with an async callback wouldn't wait for the
+  // awaits to resolve, and we'd return an empty result
```
I understand about not wanting to return an empty object, but do we need to do these serially? I believe we can use `Promise.all` to run them in parallel and `await` the results, but I'll wait until it's a PR so I can test locally.

skh replied:
Good point, changed.

jfsiii commented on x-pack/legacy/plugins/integrations_manager/common/types.ts:
```diff
@@ -10,14 +10,14 @@ export { Request, ResponseToolkit, Server, ServerRoute } from 'hapi';
-  | 'search'
-  | 'timelion-sheet'
-  | 'visualization';
+export enum AssetType {
```
I hope this works. It's what I wanted to do initially, but had some challenges. I think I've both fixed the things that were getting in the way and have a better hand on TS but I'll run locally to confirm. 🤞

jfsiii replied:
Ran locally and things still work as expected. I was struggling with a way to have these strings as both a union type & some way to do "lookup" for the "does type x use saved objects?" question. I ended up using a `Set` for `SAVED_OBJECT_TYPES` and we can create that from a few different structures (this PR shows that) so seems ok.

Can you talk about why you changed it to an `enum`? Is it solely to be able to use code like `AssetType.config` (and have one linked property name vs N duplicated strings) or are there other benefits?

FWIW, I think I'm ok changing to this. I'm just documenting what I can remember about why I went one way, and I'd like to know more about the benefits and trade-offs. I'm still new to TS & enum so I don't have practical experience with any benefits/drawbacks.

skh replied:
I wanted to be able to iterate over them, even if in the end I didn't need to. Also, for me it feels like the correct type -- we use them in the metrics UI for things like node types so I got used to using them there.

jfsiii commented on x-pack/legacy/plugins/integrations_manager/server/packages/handlers.ts:
```diff
@@ -27,24 +27,19 @@ interface ListPackagesRequest extends Request {
-
-interface DeleteAssetRequest extends Request {
-  params: AssetRequestParams;
+interface InstallDeletePackageRequest extends Request {
```
I liked that these were separate and am 50/50 on the name but we're not exporting and we can always split them up later.

skh replied:
Yeah, the important change here is `Asset` -> `Package`. I can split them, but I couldn't think (for now) of a situation where they would be different.

jfsiii commented on x-pack/legacy/plugins/integrations_manager/server/packages/install.ts:
```diff
@@ -8,35 +8,40 @@ import { SavedObject, SavedObjectsClientContract } from 'src/core/server/';
+  // Only install Kibana assets during package installation. All other asset types need special handling
+  const assetTypes: AssetType[] = [AssetType.visualization, AssetType.dashboard, AssetType.search];
+
+  let result = {};
```
I'm curious about the new return value. I'll run locally, but this is a reminder it'd be nice to have some tests to demonstrate/confirm the behavior.

skh replied:
I've rewritten this a bit and now call `getPackageInfo()` after both installation and deletion of a package. It's another `await`, but at this point we're sure all installation steps are done, and the information contained in this response may even be more useful than just the installed objects array.

jfsiii commented on x-pack/legacy/plugins/integrations_manager/server/packages/remove.ts:
```diff
@@ -21,11 +23,22 @@ export async function removeInstallation(options: {
   return installedObjects;
 }
+
+async function deletePipeline(callCluster: CallESAsCurrentUser, id: string): Promise<any> {
```
I think this is
```suggestion
async function deletePipeline(callCluster: CallESAsCurrentUser, id: string): Promise<void> {
```

because never return a value

skh replied:
Good catch, thank you!

skh commented:
> CI is failing because timelion-sheet was removed as an AssetType in two of three referenced places. If the removal was intentional, can you add a note about why? It was included because it's listed as a type in the Saved Objects API docs

I removed it because there's no mention of timelion in https://github.com/elastic/package-registry/blob/master/ASSETS.md . I've added it back in, because either source of truth is good.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/8060/)
* Commit: 6554a240f0dfc50d68b739af20b93fcca61566b5

skh commented:
@jfsiii Thank you for your comments! I've made a few changes accordingly, this should be ready for another look.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/8071/)
* Commit: dd1ce825737336894c0c7ff1067d708266c50eed

jfsiii commented:
@skh I'd like to keep the asset-specific logic in `installAssets`. 

What do you think about this? Basically, I moved kept the install rule changes in `installAssets` and kept `installPackage` the same as it is in the feature branch.

I removed some `await`s that weren't needed if we don't need the actual value for a later line

```diff
diff --git a/x-pack/legacy/plugins/integrations_manager/server/packages/install.ts b/x-pack/legacy/plugins/integrations_manager/server/packages/install.ts
index 90fad8fee4..7aec740846 100644
--- a/x-pack/legacy/plugins/integrations_manager/server/packages/install.ts
+++ b/x-pack/legacy/plugins/integrations_manager/server/packages/install.ts
@@ -8,7 +8,7 @@ import { SavedObject, SavedObjectsClientContract } from 'src/core/server/';
 import { SAVED_OBJECT_TYPE } from '../../common/constants';
 import { AssetReference, AssetType, InstallationAttributes } from '../../common/types';
 import * as Registry from '../registry';
-import { assetUsesObjects, getInstallationObject, getPackageInfo } from './index';
+import { getInstallationObject, getPackageInfo } from './index';
 import { getObjects } from './get_objects';
 
 export async function installPackage(options: {
@@ -17,31 +17,21 @@ export async function installPackage(options: {
 }) {
   const { savedObjectsClient, pkgkey } = options;
 
-  // Only install certain Kibana assets during package installation.
-  // All other asset types need special handling
-  const assetTypes: AssetType[] = [AssetType.visualization, AssetType.dashboard, AssetType.search];
+  const toSave = await installAssets({
+    savedObjectsClient,
+    pkgkey,
+  });
 
-  const installPromises = assetTypes.map(async assetType => {
-    const toSave = await installAssets({
+  if (toSave.length) {
+    // Save those references in the integration manager's state saved object
+    await saveInstallationReferences({
       savedObjectsClient,
       pkgkey,
-      assetType,
+      toSave,
     });
+  }
 
-    if (toSave.length) {
-      // Save those references in the integration manager's state saved object
-      await saveInstallationReferences({
-        savedObjectsClient,
-        pkgkey,
-        toSave,
-      });
-    }
-  });
-  await Promise.all(installPromises);
-
-  const packageInfo = await getPackageInfo({ savedObjectsClient, pkgkey });
-
-  return packageInfo;
+  return getPackageInfo({ savedObjectsClient, pkgkey });
 }
 
 // the function which how to install each of the various asset types
@@ -50,15 +40,20 @@ export async function installPackage(options: {
 export async function installAssets(options: {
   savedObjectsClient: SavedObjectsClientContract;
   pkgkey: string;
-  assetType: AssetType;
 }) {
-  const { savedObjectsClient, pkgkey, assetType } = options;
-  if (assetUsesObjects(assetType)) {
-    const references = await installKibanaSavedObjects({ savedObjectsClient, pkgkey, assetType });
-    return references;
-  }
+  const { savedObjectsClient, pkgkey } = options;
+
+  // Only install certain Kibana assets during package installation.
+  // All other asset types need special handling
+  const typesToInstall = [AssetType.visualization, AssetType.dashboard, AssetType.search];
+
+  const installationPromises = typesToInstall.map(async assetType =>
+    installKibanaSavedObjects({ savedObjectsClient, pkgkey, assetType })
+  );
 
-  return [];
+  // installKibanaSavedObjects returns AssetReference[], so .map creates AssetReference[][]
+  // call .flat to flatten into one dimensional array
+  return Promise.all(installationPromises).then(results => results.flat());
 }
 
 export async function saveInstallationReferences(options: {
```

skh commented:
@jfsiii Sure. Can you push that to my branch?

skh commented:
> Can you push that to my branch?
There's no need for that, I'm already looking into it.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/8432/)
* Commit: 5a1b73c2e3eefdd32c23c8da8040ae1b82f958f7

skh commented:
That CI failure is from https://github.com/elastic/kibana/issues/50005

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/8443/)
* Commit: bfe6a218b36c10cc83179f7d95bd800324444cd6

ruflin commented on x-pack/legacy/plugins/integrations_manager/common/types.ts:
```diff
@@ -10,14 +10,15 @@ export { Request, ResponseToolkit, Server, ServerRoute } from 'hapi';
+  search = 'search',
+  ingestPipeline = 'ingest-pipeline',
+  indexPattern = 'index-pattern',
+  timelionSheet = 'timelion-sheet',
```
Not for this PR but more general thought on this: The reason I prefixed the asset always by the service is to prevent conflicts in the future and make them unique. So if there is an `index-pattern` in Kibana and Elasticsearch, both is possible. Because of this I think we should start to refer to either `elasticsearch/ingest-pipeline` or have an asset type as a combination of the service and type.

skh replied:
Good point. At the moment some of these double as Kibana saved-object types (as defined here: https://www.elastic.co/guide/en/kibana/master/saved-objects-api-get.html) so we'll probably want to uncouple our asset types from those and only map between them.

elasticmachine commented:
Pinging @elastic/epm (Feature:EPM)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/9018/)
* Commit: efef7405b6a160184d47d4222e9fe39423a0a198

jfsiii commented on x-pack/legacy/plugins/integrations_manager/server/packages/install.ts:
```diff
@@ -45,20 +40,20 @@ export async function installPackage(options: {
+
+  // Only install certain Kibana assets during package installation.
+  // All other asset types need special handling
+  const typesToInstall = [AssetType.visualization, AssetType.dashboard, AssetType.search];
```
Are these 3 different than the other `SAVED_OBJECT_TYPES`? Could this be 
```suggestion
  const typesToInstall = Array.from(SAVED_OBJECT_TYPES);
```

skh replied:
Yes, they are different. I only included those Kibana assets here of which I'm more or less sure, for now, that we can just copy them over.

Index patterns will need to be generated from `fields.yml` (we have tasks for that), and the other Kibana asset types are not listed in https://github.com/elastic/package-registry/blob/master/ASSETS.md so it's not specified what to do with them.

jfsiii commented on x-pack/legacy/plugins/integrations_manager/server/packages/remove.ts:
```diff
@@ -21,11 +28,24 @@ export async function removeInstallation(options: {
+}
+
+async function deletePipeline(callCluster: CallESAsCurrentUser, id: string): Promise<void> {
+  // '*' shouldn't ever appear here, but it still would delete all ingest pipelines
```
Should we do something (log, error, etc) when we get an unexpected value?

skh replied:
I think we should. I'll open an issue.
